File: Code_archive/gtask4/gtask4/Clip.cpp
#include "stdafx.h"
#include "Transform.h"
#include "Clip.h"

bool Clip(point &A, point &B, point Pmin, point Pmax)
{
	bool visiable = true;
	float tMin = 0.0;
	float tMax = 1.0;

	float N[4][2] = { { 1,0 },{ 0,-1 },{ -1,0 },{ 0,1 } };
	float F[4][2] = { { Pmin.x, Pmin.y },{ Pmin.x, Pmax.y },{ Pmax.x, Pmax.y },{ Pmax.x, Pmin.y } };

	float p[2] = { B.x - A.x,B.y - A.y };
	float p1[2] = { A.x,A.y };

	for (int i = 1; i <= 4; i++)
	{
		float P = p[0] * N[i - 1][0] + p[1] * N[i - 1][1];
		float Q = (p1[0] - F[i - 1][0])*N[i - 1][0] + (p1[1] - F[i - 1][1])*N[i - 1][1];
		if (P == 0)
		{
			if (Q < 0)
			{
				visiable = false;
				break;
			}
		}
		else
		{
			if (P > 0)
			{
				tMin = System::Math::Max(tMin, -Q / P);
			}
			if (P < 0)
			{
				tMax = System::Math::Min(tMax, -Q / P);
			}
		}
		if (tMin > tMax)
		{
			visiable = false;
			break;
		}
	}
	if (visiable)
	{
		float x1 = A.x + (B.x - A.x) * tMin;
		float x2 = A.x + (B.x - A.x) * tMax;;
		float y1 = A.y + (B.y - A.y) * tMin;
		float y2 = A.y + (B.y - A.y) * tMax;;

		A.x = x1;
		B.x = x2;
		A.y = y1;
		B.y = y2;

		return true;
	}

	return false;
}

File: Code_archive/Компьютерная графика/3/3/Clip.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace _3
{
    class Clip
    {
        double scalar(ref Point A, ref Point B)
        {
            return A._x * B._x + A._y * B._y;
        }

        public bool clip(ref Point A, ref Point B, Point Pmin, Point Pmax)
        {
            float tmin = 0, tmax = 1;
            int pi = 0, qi = 0;
            for (int i = 1; i < 5; i++)
            {
                Point p = new Point();
                p._x = B._x - A._x;
                p._y = B._y - A._y;
                Point N = new Point();
                N._x = 0;
                N._y = 0;
                Point F = new Point();

                switch (i)
                {
                    case 1:
                        N._x = 1;
                        F = Pmin;
                        break;
                    case 2:
                        N._y = -1;
                        F._x = Pmin._x;
                        F._y = Pmax._y;
                        break;
                    case 3:
                        N._x = -1;
                        F = Pmax;
                        break;
                    case 4:
                        N._y = 1;
                        F._x = Pmax._x;
                        F._y = Pmin._y;
                        break;
                }
                pi = (int)scalar(ref p, ref N);
                Point sub = new Point();
                sub._x = A._x - F._x;
                sub._y = A._y - F._y;
                qi = (int)scalar(ref sub, ref N);

                if (pi == 0)
                {
                    if (qi < 0)
                        return false;
                }
                else
                {
                    if (pi > 0)
                        tmin = Math.Max(tmin, -(float)qi / pi);
                    else
                        tmax = Math.Min(tmax, -(float)qi / pi);
                    if (tmin > tmax)
                        return false;
                }
            }

            Point An = new Point();
            Point Bn = new Point();
            An._x = A._x + (B._x - A._x) * tmin;
            An._y = A._y + (B._y - A._y) * tmin;
            Bn._x = A._x + (B._x - A._x) * tmax;
            Bn._y = A._y + (B._y - A._y) * tmax;
            A = An;
            B = Bn;
            return true;
        }
    }
}
